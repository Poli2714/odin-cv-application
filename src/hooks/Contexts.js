import { createContext } from 'react';

const IndexContext = createContext({
  activeIndex: null,
  setActiveIndex: () => {},
});

const KeyContext = createContext({
  activeKey: {
    eduActiveKey: null,
    expActiveKey: null,
  },
  setActiveKey: () => {},
});

const EduErrMsgContext = createContext({
  showEduErrMessage: false,
  setShowEduErrMessage: () => {},
});

const ExpErrMsgContext = createContext({
  showExpErrMessage: false,
  setShowExpErrMessage: () => {},
});

const PersonalInfoContext = createContext({
  personalInfo: {
    profilePicture: '',
    fullName: '',
    currentJobTitle: '',
    email: '',
    mobile: '',
    city: '',
  },
  setPersonalInfo: () => {},
});

const SummaryContext = createContext({
  summary: '',
  setSummary: () => {},
});

const ExperienceContext = createContext({
  experiences: {
    id: {
      jobTitle: '',
      employer: '',
      jobStartDate: '',
      jobEndDate: '',
      jobDescription: '',
    },
  },
  setExperiences: () => {},
});

const EducationContext = createContext({
  educations: {
    id: {
      degree: '',
      university: '',
      eduStartDate: '',
      eduEndDate: '',
      eduInfo: '',
    },
  },
  setEducations: () => {},
});

export {
  EducationContext,
  EduErrMsgContext,
  ExperienceContext,
  ExpErrMsgContext,
  IndexContext,
  KeyContext,
  PersonalInfoContext,
  SummaryContext,
};
