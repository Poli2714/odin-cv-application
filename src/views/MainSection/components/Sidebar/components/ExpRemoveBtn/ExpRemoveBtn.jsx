import { useContext } from 'react';

import { RemoveBtn } from 'src/components/buttons';
import { ExperienceContext, KeyContext } from 'src/hooks/Contexts';

function ExpRemoveBtn() {
  const { experiences, setExperiences } = useContext(ExperienceContext);
  const { activeKey, setActiveKey } = useContext(KeyContext);
  const { expActiveKey } = activeKey;

  const handleClickRemoveExp = key => {
    const newExperiences = { ...experiences };
    delete newExperiences[key];
    setExperiences(newExperiences);
    setActiveKey({
      ...activeKey,
      expActiveKey: Object.keys(newExperiences).at(-1),
    });
  };

  return (
    <RemoveBtn
      disabled={Object.keys(experiences).length === 1 || !expActiveKey}
      onClick={() => handleClickRemoveExp(expActiveKey)}
    />
  );
}

export default ExpRemoveBtn;
