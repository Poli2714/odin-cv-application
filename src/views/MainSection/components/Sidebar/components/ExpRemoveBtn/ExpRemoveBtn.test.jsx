import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';

import ExpRemoveBtn from './ExpRemoveBtn';
import { ExperienceContext, KeyContext } from 'src/hooks/Contexts';

vi.mock('src/components/buttons', () => ({
  RemoveBtn: vi.fn(({ disabled }) => (
    <button disabled={disabled}>Mock click</button>
  )),
}));

function MockParentComponent({ activeKey, experiences }) {
  return (
    <ExperienceContext.Provider value={{ experiences }}>
      <KeyContext.Provider value={{ activeKey }}>
        <ExpRemoveBtn />
      </KeyContext.Provider>
    </ExperienceContext.Provider>
  );
}

test('renders ExpRemoveBtn', () => {
  render(<ExpRemoveBtn />);

  expect(screen.getByRole('button')).toMatchSnapshot();
});

test('disables RemoveBtn if there is only one experience', async () => {
  const experiences = { id: { someData: '' } };
  const activeKey = { expActiveKey: 'key' };

  render(
    <MockParentComponent activeKey={activeKey} experiences={experiences} />,
  );
  const button = screen.getByRole('button');

  expect(button).toBeDisabled();
});

test('disables RemoveBtn if there are more than one experience and all are collapsed (not displayed)', () => {
  const experiences = {
    id: { someData: 'foo' },
    id2: { someOtherData: 'bar' },
  };
  const activeKey = { expActiveKey: '' };

  render(
    <MockParentComponent activeKey={activeKey} experiences={experiences} />,
  );
  const button = screen.getByRole('button');

  expect(button).toBeDisabled();
});

test('enables RemoveBtn if there are more than one experience and one of them is expanded (displayed)', () => {
  const experiences = {
    id: { someData: 'foo' },
    id2: { someOtherData: 'bar' },
  };
  const activeKey = { expActiveKey: 'hey' };

  render(
    <MockParentComponent activeKey={activeKey} experiences={experiences} />,
  );
  const button = screen.getByRole('button');

  expect(button).toBeEnabled();
});
