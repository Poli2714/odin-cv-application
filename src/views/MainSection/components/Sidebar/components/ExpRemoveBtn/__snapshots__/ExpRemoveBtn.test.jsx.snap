// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders ExpRemoveBtn 1`] = `
<button
  disabled=""
>
  Mock click
</button>
`;
